### PR TITLE
Add few unsafe markers

### LINKED ISSUE #2453: Mark most functions as unsafe
Because we follow Vulkan closely for the HAL API, we don't have many safety guarantees for most of the API usage. A few users have asked to have the explicit unsafe annotation to make it more obvious that no guarantees are provided here.

### PR DESCRIPTION
**Pull request by omni-viral**
Fixes #2453
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends:
  metal
- [ ] `rustfmt` run on changed code

### CODE REVIEW

File: examples/Cargo.toml [hunk 1]
```diff
@@ -3,6 +3,7 @@ name = "hal-examples"
 version = "0.1.0"
 publish = false
 workspace = ".."
+edition = "2018"
```
Comment (kvark) [inline examples/Cargo.toml]: let's leave the edition update to other PRs, the scope is already high, unless you require it for this to work

Comment (omni-viral) [reply examples/Cargo.toml]: The code needs nll.

**Comment by kvark** [comment]
Can you remove the second commit please?

**Comment by kvark** [comment]
The issue https://github.com/rust-lang/rust/issues/57111 is resolved now, please remove the last commit.

**Comment by omni-viral** [comment]
If you insist

**Comment by kvark** [comment]
Thank you!
Bors r+

**Pull request merged**